src/Microsoft.VisualStudio.Threading/AsyncReaderWriterLock.cs

```diff
@@ -2430,6 +2429,47 @@ internal void SetFault(Exception ex)
                 this.fault = ex;
             }
 
+            /// <summary>
+            /// Get the correct SynchronizationContext to execute code executing within the lock.
+            /// </summary>
+            private SynchronizationContext GetEffectiveSynchronizationContext()
+            {
+                if (this.synchronizationContext == null)
+                {
+                    // Only read locks can be executed trivially. The locks that have some level of exclusivity (upgradeable read and write)
+                    // must be executed via the NonConcurrentSynchronizationContext.
+                    SynchronizationContext synchronizationContext = null;
+
+                    Awaiter awaiter = this.NestingLock;
+                    while (awaiter != null)
+                    {
+                        if (this.lck.IsLockActive(awaiter, considerStaActive: true))
```

Comment:
Why are we checking whether the nesting lock is active, but we don't check whether our own lock is active? A code comment explaining why our own `this.lck.IsLockActive(this)` isn't worthwhile would be good to add.

Reply:
It looks like your goal is to share the SynchronizationContext created by any ancestor lock to avoid creating a new one here. That makes sense. And might be that code comment that I was looking for. :smile:

Reply:
Yes, added comments.